### ISSUE TITLE
OpenCV face detector speed is unstable on 1920x1080 frame resized to 300x300 before network input

### ISSUE THREAD

**Issue by fortminors** [opencv/opencv#20730]
##### System information (version)

- OpenCV => 4.5.3 (Compiled from source with CUDA support)
- Operating System / Platform => Windows 10 64 Bit
- Compiler => Visual Studio 2019

##### Detailed description

I am investigating the face detectors that OpenCV offers. I tried both caffe and tensorflow models and it works great when the camera input frame is of 640x480 resolution. The network runs in about 7ms for every frame. However, when I try to set the input from camera to 1920x1080 resolution and then resizing to 640x480 or 300x300 using the OpenCV resize function and feeding it into the network, the running time starts to jump around from 6ms to 25ms (see the screenshot below), which is quite weird. I don't understand how can this behaviour be caused, as the input to the network is always a 300x300 frame...

![scrn](https://user-images.githubusercontent.com/36015059/134334399-c7dbf1b4-db9d-4206-bf65-4128357eb09c.png)

The files required for the models are [here](https://drive.google.com/file/d/1rV8f59nSnOXoyBF7498KJyKjvJZRwzY2/view?usp=sharing).

What is more interesting, is that if I set `DNN_TARGET_CPU` for the network, the stability is returned even on the 1920x1080 image without resizing for 13ms per frame.

##### Steps to reproduce (C++)

```
#include <iostream>
#include <string>
#include <vector>
#include <stdlib.h>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/dnn.hpp>

using namespace cv;
using namespace cv::dnn;
using namespace std;

const size_t inWidth = 300;
const size_t inHeight = 300;
const double inScaleFactor = 1.0;
const float confidenceThreshold = 0.7;
const cv::Scalar meanVal(104.0, 117.0, 123.0);

const std::string caffeConfigFile = "models/deploy.prototxt";
const std::string caffeWeightFile = "models/res10_300x300_ssd_iter_140000_fp16.caffemodel";

const std::string tensorflowConfigFile = "models/opencv_face_detector.pbtxt";
const std::string tensorflowWeightFile = "models/opencv_face_detector_uint8.pb";

using std::chrono::high_resolution_clock;
using std::chrono::duration_cast;
using std::chrono::duration;
using std::chrono::milliseconds;

void detectFaceOpenCVDNN(Net net, Mat& frameOpenCVDNN, string framework)
{
    int frameHeight = frameOpenCVDNN.rows;
    int frameWidth = frameOpenCVDNN.cols;

    cv::Mat inputBlob;
    if (framework == "caffe")
        inputBlob = cv::dnn::blobFromImage(frameOpenCVDNN, inScaleFactor, cv::Size(inWidth, inHeight), meanVal, false, false);
    else
        inputBlob = cv::dnn::blobFromImage(frameOpenCVDNN, inScaleFactor, cv::Size(inWidth, inHeight), meanVal, true, false);

    net.setInput(inputBlob, "data");
    cv::Mat detection = net.forward("detection_out");

    cv::Mat detectionMat(detection.size[2], detection.size[3], CV_32F, detection.ptr<float>());

    for (int i = 0; i < detectionMat.rows; i++)
    {
        float confidence = detectionMat.at<float>(i, 2);

        if (confidence > confidenceThreshold)
        {
            int x1 = static_cast<int>(detectionMat.at<float>(i, 3) * frameWidth);
            int y1 = static_cast<int>(detectionMat.at<float>(i, 4) * frameHeight);
            int x2 = static_cast<int>(detectionMat.at<float>(i, 5) * frameWidth);
            int y2 = static_cast<int>(detectionMat.at<float>(i, 6) * frameHeight);

            cv::rectangle(frameOpenCVDNN, cv::Point(x1, y1), cv::Point(x2, y2), cv::Scalar(0, 255, 0), 2, 4);
        }
    }

}


int main()
{
    string videoFileName;
    string device;
    string framework;

    device = "GPU";
    framework = "caffe";

    cout << "Configuration:" << endl;
    cout << "Device - " << device << endl;
    if (framework == "caffe")
        cout << "Network type - Caffe" << endl;
    else
        cout << "Network type - TensorFlow" << endl;
    if (videoFileName == "")
        cout << "No video found, using camera stream" << endl;
    else
        cout << "Video file - " << videoFileName << endl;

    Net net;

    if (framework == "caffe")
        net = cv::dnn::readNetFromCaffe(caffeConfigFile, caffeWeightFile);
    else
        net = cv::dnn::readNetFromTensorflow(tensorflowWeightFile, tensorflowConfigFile);

#if (CV_MAJOR_VERSION >= 4)
    if (device == "CPU")
    {
        net.setPreferableBackend(DNN_TARGET_CPU);
    }
    else
    {
        net.setPreferableBackend(DNN_BACKEND_CUDA);
        net.setPreferableTarget(DNN_TARGET_CUDA);
    }
#else
    // force CPU backend for OpenCV 3.x as CUDA backend is not supported there
    net.setPreferableBackend(DNN_BACKEND_DEFAULT);
    device = "cpu";
#endif

    cv::VideoCapture source(0);// , CAP_V4L);

    source.set(cv::CAP_PROP_FRAME_WIDTH, 1920);
    source.set(cv::CAP_PROP_FRAME_HEIGHT, 1080);

    Mat frame;

    double tt_opencvDNN = 0;
    double fpsOpencvDNN = 0;

    while (true)
    {
        source >> frame;
        if (frame.empty())
        {
            break;
        }

        double t = cv::getTickCount();

        auto t1 = high_resolution_clock::now();

        // Remove this and set the source width and height to 640, 480 respectively
        // And observe how the running time drops to stable 7ms
        Mat resizedFrame;
        resize(frame, resizedFrame, Size(inWidth, inHeight));
        detectFaceOpenCVDNN(net, resizedFrame, framework);

        //detectFaceOpenCVDNN(net, frame, framework);

        auto t2 = high_resolution_clock::now();

        auto time = duration_cast<milliseconds>(t2 - t1);

        cout << "Time taken = " << time.count() << " ms" << endl;

        tt_opencvDNN = ((double)cv::getTickCount() - t) / cv::getTickFrequency();
        fpsOpencvDNN = 1 / tt_opencvDNN;

        putText(frame, format("OpenCV DNN %s FPS = %.2f", device.c_str(), fpsOpencvDNN), Point(10, 50), FONT_HERSHEY_SIMPLEX, 1.3, Scalar(0, 0, 255), 4);

        imshow("OpenCV - DNN Face Detection", frame);

        int k = waitKey(5);
        if (k == 27)
        {
            source.release();
            destroyAllWindows();
            break;
        }
    }
}
```

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by YashasSamaga** [comment]
I could not reproduce this on Ubuntu 20.04. I don't have a Windows setup to run OpenCV at the moment. But I have a hypothesis that you can test.

I suspect this has something to do with how Windows schedules processes. The CUDA backend execution is asynchronous: work is dumped to the GPU and the main CPU thread waits for the work to finish. The waiting procedure as of now is a spinlock which may help improve small benchmarks slightly (or negligibly) but is probably a terrible thing to do in practice. This might make the OS scheduler treat your process as a compute-intensive job/background task instead of an interactive process. This in turn might reduce the odds of your process being scheduled again quickly after it's pre-empted. It's a guess. You can confirm using a profiler.

If that is indeed the case, you might be able to fix it with this change:

Apply the diff below on this:

https://github.com/opencv/opencv/blob/38b9ec7a18e95140a1f34d7b1c52f4fe74c14459/modules/dnn/src/cuda4dnn/csl/event.hpp#L33-L39

```diff
--- a/modules/dnn/src/cuda4dnn/csl/event.hpp
+++ b/modules/dnn/src/cuda4dnn/csl/event.hpp
@@ -34,7 +34,7 @@ namespace cv { namespace dnn { namespace cuda4dnn { namespace csl {
         Event(bool create, bool timing_event = false) : event{nullptr} {
             if (create) {
                 unsigned int flags = (timing_event ? 0 : cudaEventDisableTiming);
-                CUDA4DNN_CHECK_CUDA(cudaEventCreateWithFlags(&event, flags));
+                CUDA4DNN_CHECK_CUDA(cudaEventCreateWithFlags(&event, flags | cudaEventBlockingSync));
             }
         }
```

Please let me know if you try it. I think the synchronization waiting for GPU results should be a blocking wait instead of a spinlock.

**Comment by YashasSamaga** [comment]
I just realized that you're timings include the resize call too. Can you please try measuring the time `net.forward()` takes? Maybe the stochastic performance behavior isn't arising out of the DNN.

You're also allocating a new `Mat` every iteration. Can you try moving `resizedFrame` outside?

**Comment by fortminors** [comment]
@YashasSamaga thank you for such a comprehensive reply.

I should've mentioned that I actually measured time just for the 'net.forward' call but moved it outside the function as everything else was taking 0 ms.

Regarding the spinlock, I didn't know this detail, but what you say makes sense.

When I have time, I will check out your suggestion and write back here, thanks!

**Comment by fortminors** [comment]
@YashasSamaga Interestingly, I tried to reproduce the problem today before trying out your suggestions, but it worked well with 7 ms per frame on CUDA backend. I changed however the camera that I've used. I'm on Logitech BRIO atm and it works great. I don't remember if I changed any code, but I will post the one that I have not anyways. Thanks a lot!

Code:

```
#include <iostream>
#include <string>
#include <vector>
#include <stdlib.h>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/face.hpp>
#include <opencv2/opencv.hpp>

using namespace cv;
using namespace cv::dnn;
using namespace std;
using namespace cv::face;

const size_t inWidth = 300;
const size_t inHeight = 300;
const double inScaleFactor = 1.0;
const float confidenceThreshold = 0.7;
const cv::Scalar meanVal(104.0, 117.0, 123.0);

const std::string caffeConfigFile = "models/deploy.prototxt";
const std::string caffeWeightFile = "models/res10_300x300_ssd_iter_140000_fp16.caffemodel";

const std::string tensorflowConfigFile = "models/opencv_face_detector.pbtxt";
const std::string tensorflowWeightFile = "models/opencv_face_detector_uint8.pb";

const string landmarkModel = "shape_predictor_68_face_landmarks.dat";

using std::chrono::high_resolution_clock;
using std::chrono::duration_cast;
using std::chrono::duration;
using std::chrono::milliseconds;

std::vector<Rect> DetectFaces(Net net, Mat& frameOpenCVDNN)
{
    int frameHeight = frameOpenCVDNN.rows;
    int frameWidth = frameOpenCVDNN.cols;

    cv::Mat inputBlob;
    inputBlob = cv::dnn::blobFromImage(frameOpenCVDNN, inScaleFactor, cv::Size(inWidth, inHeight), meanVal, false, false);

    net.setInput(inputBlob, "data");

    auto t1 = high_resolution_clock::now();

    Mat detection = net.forward("detection_out");

    auto t2 = high_resolution_clock::now();

    auto time = duration_cast<milliseconds>(t2 - t1);

    cout << "Detect time = " << time.count() << endl;

    Mat detectionMat(detection.size[2], detection.size[3], CV_32F, detection.ptr<float>());

    std::vector<Rect> faces;

    for (int i = 0; i < detectionMat.rows; i++)
    {
        float confidence = detectionMat.at<float>(i, 2);

        if (confidence > confidenceThreshold)
        {
            int x1 = static_cast<int>(detectionMat.at<float>(i, 3) * frameWidth);
            int y1 = static_cast<int>(detectionMat.at<float>(i, 4) * frameHeight);
            int x2 = static_cast<int>(detectionMat.at<float>(i, 5) * frameWidth);
            int y2 = static_cast<int>(detectionMat.at<float>(i, 6) * frameHeight);

            //cv::rectangle(frameOpenCVDNN, cv::Point(x1, y1), cv::Point(x2, y2), cv::Scalar(0, 255, 0), 2, 4);

            faces.emplace_back(x1, y1, (x2 - x1), (y2 - y1));
        }
    }

    return faces;
}

int main()
{
    Net net;

    net = cv::dnn::readNetFromCaffe(caffeConfigFile, caffeWeightFile);

    //net.setPreferableBackend(DNN_BACKEND_DEFAULT);
    //net.setPreferableTarget(DNN_TARGET_CPU);

    net.setPreferableBackend(DNN_BACKEND_CUDA);
    net.setPreferableTarget(DNN_TARGET_CUDA);

    cv::VideoCapture source(0);// , CAP_V4L);

    source.set(cv::CAP_PROP_FRAME_WIDTH, 1920);
    source.set(cv::CAP_PROP_FRAME_HEIGHT, 1080);

    Mat frame;

    double tt_opencvDNN = 0;
    double fpsOpencvDNN = 0;

    while (true)
    {
        source >> frame;
        if (frame.empty())
        {
            break;
        }

        double t = cv::getTickCount();

        std::vector<Rect> faces = DetectFaces(net, frame);

        tt_opencvDNN = ((double)cv::getTickCount() - t) / cv::getTickFrequency();
        fpsOpencvDNN = 1 / tt_opencvDNN;

        putText(frame, format("OpenCV DNN %s FPS = %.2f", "CPU", fpsOpencvDNN), Point(10, 50), FONT_HERSHEY_SIMPLEX, 1.3, Scalar(0, 0, 255), 4);

        imshow("OpenCV - DNN Face Detection", frame);

        int k = waitKey(5);
        if (k == 113) // q
        {
            source.release();
            destroyAllWindows();
            break;
        }
    }
}

```
